Conditionally enable and disable CLI options

mabartos opened the pull request:
Closes #13113

The intention behind the lazy evaluation of `isEnabled()` for property mappers is to have the possibility to enable/disable it based on the current configuration with no need to keep the order of property mapper groups.

In that case, we can enable/disable options based on values of other options as f.e. this is enabled only when the property kc.xyz=gold, etc.

vmuzikar commented:
CC @shawkins

mabartos commented:
@shawkins This PR should be ready for review. I'd be happy if you'd have a chance to look at it. 

I still need to finish some additional things:

- [ ] Add description to this PR with the provided functionality (and possibly show some outputs)
- [ ] Add/fix some more tests
- [ ] When the distro is not augmented, the validation of the disabled runtime options is made after the reaugmentation (f.e. when `start` without `--optimized` is used; for `build` works as expected)

shawkins reviewed (commented):
Just a couple of nits in review comments.

If I understand correctly this will handle enablement as long as it's conditional on properties that are set prior to parsing - which most are. But anything that is set as a side effect of the command, such as the StartDev forceDevProfile that won't be known. Is that correct?

I haven't had a chance to run this, but was wondering will this make both help and help-all conditional? That is if you have something like --log=file --help-all will you see any of the file or gelf options?

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:
```diff
@@ -145,9 +154,31 @@ ConfigValue getConfigValue(String name, ConfigSourceInterceptorContext context)
+    }
+
+    public Optional<String> getEnabledWhen() {
+        return Optional.ofNullable(enabledWhen)
```
I'm not sure. I quite like this approach as it's clear on the first look that `enabledWhen` is optional and does not always have to be specified. Even though the `isNotBlank` checks the null values, we also need to check whether the string is empty or not. I tend to be a little bit more defensive in these cases to prevent NPE.

I could replace the `isNotBlank` with `.filter(f -> !f.isBlank())` to check only the string value. 

Or what do you suggest?

shawkins replied:
I'd opt for updating the utility method javadoc to specify the null case is handled - and to change it to just str != null && !str.isBlank(). Or if you want directly guard against the null case and use String.isBlank directly, then the utility method should be deprecated / removed.

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/LoggingPropertyMappers.java:
```diff
@@ -29,15 +30,21 @@ public static PropertyMapper[] getMappers() {
                         .paramLabel("<handler>")
                         .build(),
                 fromOption(LoggingOptions.LOG_CONSOLE_OUTPUT)
+                        .isEnabled(LoggingPropertyMappers::isConsoleEnabled)
```
> since these are always specified together it may make sense to just have a single method for this:

Yep, that'd be better :)

> A string contstant could be extracted for the messages, or there could even be a holder or interface that would directly associate the message with the boolean function.

Do we want to create any more comprehensive logic around it? I like the flexibility in that as there might be several different use-cases such as obtaining value from specific property `Configuration.getKcConfigValue("xyz") == "abc"` even though there might be provided some duplications. 

But if you think it'd be better, I can do it.

mabartos replied:
Or we can provide some sort of hybrid approach similar to the DeprecatedMetadata as we could potentially in the future enhance only the metadata around enablement. I'll try to write some prototype.

mabartos replied:
I tried to create EnabledMetadata instead of that, but I don't like it much. I'll merge only the builder methods for now and then we can discuss.

https://github.com/mabartos/keycloak/commit/d402d817aab28c121b68aa95f9a0a1cd833cdb70

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java:
```diff
@@ -276,10 +279,20 @@ public static void validateNonCliConfig(List<String> cliArgs, AbstractCommand ab
+            if (options.includeBuildTime) {
+                Optional.ofNullable(PropertyMappers.getDisabledBuildTimeMappers().get(category)).ifPresent(mappers::addAll);
+            }
+            if (options.includeRuntime && !isRebuildCheck()) {
```
The reason behind this is to prevent warnings about disabled runtime mappers from being displayed in the log of the JVM managing the reaugmentation. Because when the reaug is needed for `start` or `start-dev`, the `includeRuntime=true` in this case. That might be a little bit confusing, but that's the current case in Keycloak.  

I haven't investigate it more yet, but I'll take a look at the possible options.

mabartos commented:
> If I understand correctly this will handle enablement as long as it's conditional on properties that are set prior to parsing - which most are. But anything that is set as a side effect of the command, such as the StartDev forceDevProfile that won't be known. Is that correct?

It should handle such situations, but I'll write a test for it and double-check it. 

I haven't had a chance to run this, but was wondering will this make both help and help-all conditional? That is if you have something like --log=file --help-all will you see any of the file or gelf options?

> Only `--help` is conditional. If you specify `--help-all`, all options are printed, even disabled ones with the message when the option is available to the user.

vmuzikar reviewed (commented):
@mabartos Thanks for the PR! The general approach LGTM, although I haven't yet reviewed line-by-line.

I think we still need to tackle one more case here: when some option has multiple versions, essentially when we add an option with the same key multiple times. I think it doesn't work yet with this version of the PR, right? This will be required for the hostname stuff when some of the options will have new meaning (descriptions, potentially transformers etc.).

Could you also please rebase to incorporate changes from now merged #24811?

mabartos commented:
@vmuzikar You're right, I'll take it into consideration.

Anyway, I need to solve some other minor cases around disabled properties evaluation.

mabartos commented:
## Characteristics of provided changes

As this PR is quite comprehensive, it'd be good to briefly describe the provided changes and intentions. 

### General
- **Lazy evaluation of conditions** - in order to avoid maintaining the initialization order of Property mappers (we can create conditions based on the value of a particular option)
- **All options are available during the evaluation of conditions** - we need to determine which options/mappers will be disabled based on property values from all config sources (to be able assembling conditions from all options) 
- **After the evaluation of the conditions, config sources are sanitized** - after the evaluation, we do not want to know anything about the disabled options. Simply, we don't know they exist. Sanitization of properties (interface `SanitizableConfigSource`) was preferred over the re-initialization of config sources again due to performance reasons.
- **Runtime/build-time disabled options** - Disabled options are evaluated during the phase they are relevant to. Build time disabled options are evaluated in build-time and runtime options during runtime.

### Specify Disabled options
- **Directly specified via CLI** - when a disabled option (based on condition evaluation) is specified via CLI, the error message is shown. For instance for command `./kc.sh start --log=console --log-file-output-json`, the message is:
```
Disabled option: '--log-file-output'. Available only when File log handler is activated
Possible solutions: --log, --log-console-output, --log-console-format, --log-console-color, --log-level
Try 'kc.sh start --help' for more information on the available options.
Specify '--help-all' to obtain information on all options and their availability.
```
- **Specified in other sources** - when a disabled options is specified in other sources than CLI (KC properties file, ENV var, keycloak.conf,...), only warning about ignoring these properties is printed:
```
WARN [org.ke.qua.run.cli.Picocli] The following used options are UNAVAILABLE and will be ignored:
        - log-gelf-level: Available only when GELF is activated.
        - log-file: Available only when File log handler is activated.
        - log-gelf-version: Available only when GELF is activated.
``` 

### Help 
- `--help-all` prints all options (even disabled ones) 
- `--help` prints only options, which are not disabled
- **Disableable Options prints state, when they're enabled** - for instance, the suffix(`Available when...`) for the `log-console-color` option:
```
--log-console-color <true|false>
                     Enable or disable colors when logging to console. Default: false. Available
                       only when Console log handler is activated.
```

### Docs
- ** All configuration guide contains all options** - in the guide, the disableable options contain also the description suffix (`Available when...`)

### Known issues
- **No support for duplicated keys yet** - ATM, there is NO support for possibility to have two options with the same key (one will be disabled) - I'll be working on it soon
- **The `Profile` is initialized twice during augmentation** - could be done in a follow-up task 

---------------------------------------------------------

@vmuzikar @shawkins @Pepo48 
Once the tests passed, this PR will be ready to review. However, I need to add the functionality related to the duplicated keys(described in the Known issues above).

mabartos commented:
UPDATE:

In order to properly manage the disabled options, it's necessary to reinitialize `CommandLine` in `Picocli` class. I tried to measure it, and it seems the reinitialization takes only ~110ms (I expected more). 

I tried to bypass it by reusing the command line, but Picocli does not support the mutability of these options.

Another approach was to explicitly revalidate CLI options on our own, but we'd touch on the controversy around maintainability/performance. I decided to not use it as additional changes would be needed and the maintainability would decrease. See https://github.com/mabartos/keycloak/commit/89c6df1bd74f3c32451fcc67fbdc3b0baedecba0

shawkins reviewed (commented):
Just a few quick thoughts / questions.

mabartos commented on quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/ConfigurationTest.java:
```diff
@@ -158,7 +178,7 @@ public void testEnvVarPriorityOverSysProps() {
     public void testCLIPriorityOverSysProp() {
         System.setProperty("kc.spi.hostname.default.frontend-url", "http://propvar.unittest");
-        System.setProperty(CLI_ARGS, "--spi-hostname-default-frontend-url=http://cli.unittest");
+        setCliArgs("--spi-hostname-default-frontend-url=http://cli.unittest");
```
Ok, might be better.

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java:
```diff
@@ -117,17 +143,46 @@ public static boolean isSupported(PropertyMapper<?> mapper) {
+    public static Optional<PropertyMapper<?>> getDisabledMapper(String property) {
+        if (property == null) return Optional.empty();
+
+        return getDisabledMappers()
```
> like the propertymappers, should we consider storing these in a map that is keyed by all of the possible candidates (to, from, cli, env).

@shawkins Yep, we probably can. 

> Also do we want to proactively consider the case where this is not 1-1? Like how we already have multiple propertymappers for example defined off of kc.db. Or should we worry about it once it comes up later?

I assume you mean the case when property mappers use the `mapFrom` attribute and the defined property by that is disabled, right? Yep, we can consider it as well. I assume the property mapper which uses disabled `mapFrom` prop should be also disabled, as it's the parent of it, right? I can try to implement it, but I'd rather do it in a follow-up task. I'd like to provide the functionality ASAP and then we can improve some other use-cases. WDYT?

shawkins replied:
> I assume you mean the case when property mappers use the mapFrom attribute and the defined property by that is disabled, right?

Yes. db, fips-mode, log, and proxy-headers all map to multiple quarkus properties - we're not talking about disabling any of those yet, so it's more of a future concern.

>  I can try to implement it, but I'd rather do it in a follow-up task. I'd like to provide the functionality ASAP and then we can improve some other use-cases. WDYT?

That sounds good. Perhaps designating one of them as the parent, or even restructing so that PropertyMappers have child will make this clearer.

shawkins commented on quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/ConfigurationTest.java:
```diff
@@ -564,20 +584,59 @@ public void testKeystoreConfigSourcePropertyMapping() {
         ConfigProviderResolver.setInstance(null);
-        return ConfigUtils.configBuilder(true, LaunchMode.NORMAL).build();
+        // reflect real config source configuration
+        return new KeycloakConfigSourceProvider().configBuilder(ConfigUtils.emptyConfigBuilder()).build();
```
Isn't KeycloakConfigSourceProvider picked up via the ServiceLoader? Or do you see a case where it is not?

mabartos replied:
The thing is that I tried to get rid of the default `EnvConfigSource`, as we don't want to have processed Keycloak ENV vars there. For that purpose, I've created custom `SmallRyeEnvConfigSource`[1].

The reason behind that is to not process values provided via disabled Keycloak ENV vars.

But if I'm not mistaken, the `EnvConfigSource` is provided into the default config sources anyway through the SmallRyeConfigProviderResolver, and its loading via the Service loader. It means, the test case for the ENV var is working only in the test class.

But I'm not sure, if I should spend more time to have the ability to ignore Keycloak ENV vars when their mappers are disabled.

@shawkins WDYT? 

[1] https://github.com/keycloak/keycloak/pull/25333/files/323aff68e704674e2cd9fa33910d07428f744fdb#diff-3574824ab9fd51968b78e779c88ff3b7ea5688a840dbd57e285518d236b84094R37

shawkins commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/SanitizableConfigSource.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Config source with the ability to sanitize its properties based on environment setup
+ */
+public interface SanitizableConfigSource {
```
> The required state of the app is that after disabling particular props, we should consider them as such as they have never existed.

This is so we can skip the validation of those values, and also not persist them - without introducing new calls into the respective logic to check for options being disabled correct?

> With the filtering, they should still exist somewhere, and if we call directly the config source, they'd be present there.

That's currently needed for the Picocli.validateConfig in some form - with the sanitize approach via KeycloakConfigSourceProvider.getSanitizedRemovedProperties(). There'd be something similar on the interceptor, or the interceptor would be temporarily disabled like PropertyMappingInterceptor.

Does your concern about the interceptor stem from other cases where you see config consumers by-passing top-level config calls and going directly against individual config sources?

> @shawkins I'm kinda on top of a fence and not sure, what's the most suitable approach. I'd probably prefer to keep the current approach as we can save some time on it. And we can do some amendments in follow-up tasks. WDYT?

Let's loop in @radcortez as well.

mabartos replied:
> This is so we can skip the validation of those values, and also not persist them - without introducing new calls into the respective logic to check for options being disabled correct?

This is a little bit tricky. We need to distinguish between 2 phases of the initialization:
1. **BEFORE** the evaluation of disabled mappers - we need to have properly initialized all config sources as normal; to be able to determine, whether the mapper will be disabled or not. 
2. **AFTER** the evaluation of disabled mappers - return values of only relevant mappers

To explain my intention more - in the first prototype, I put the check of disabled mappers in relevant places. However, I've decided not to use it as such places might be a lot and it could potentially cause some undesirable behavior. So I decided to sanitize our config sources for good after the evaluation of disabled mappers.

Did it answer your question, or you meant sth else?

radcortez replied:
> Does your concern about the interceptor stem from other cases where you see config consumers by-passing top-level config calls and going directly against individual config sources?

If you need to omit a particular property an interceptor is the best alternative. Interceptors are executed before sources are queried and are decoupled from sources, so you can add or remove behavior without changing the sources themselves.

For the interceptor you only need to filter the property being queried in `getValue` and `iterateNames`. Now, if someone gets a reference to the specific source, they could still retrieve the filtered names, but usually, this is not something that users should do.

mabartos replied:
@radcortez Thanks for your input. 

Just a few questions:
 - So, we don't need to override the `iterateValues`, is it correct?
 -  If the interceptor in extensions does not have any `@Priority` annotation, what is the default priority?
 - Is it `Priorities.APPLICATION=5000` ? (and potentially dynamically handled conflicts on the SmallRye side?)

radcortez replied:
> * So, we don't need to override the `iterateValues`, is it correct?

No. We don't use it, it is deprecated and we will remove it soon.

> * If the interceptor in extensions does not have any `@Priority` annotation, what is the default priority?
Its `Priorities#APPLICATION=5000`

> * Is it `Priorities.APPLICATION=5000` ? (and potentially dynamically handled conflicts on the SmallRye side?)
Yes. If multiple interceptors are found with the same priority, the behavior is nondeterministic. All the interceptors are added to the chain, but we don't know the execution order (for the ones sharing the priority)

mabartos replied:
@radcortez Ok, thanks

mabartos commented:
### Summarize potential solution after the dicussion above
@shawkins It'd be good to clarify the proposed approach in order to be on the same page. 

Proposed approach:

- No `SanitizableConfigSources`
- New `ConfigSourceInterceptor` with the ability to be turned **ON/OFF**
- The CS interceptor would **filter out all disabled properties**
- The CS interceptor would be turned **OFF** when checking what property mappers should be disabled (1st phase) - it means, we have access to all properties
- The CS interceptor would be turned **ON** after obtaining all disabled property mappers (2nd phase)
- The CS interceptor would be turned **OFF** when executing the `Picocli.validateConfig()` - check the disabled mappers there and be able **to properly print logs** 

I think it might be a good approach. 

@shawkins Does it work for you? Or what would you change?

shawkins commented:
> @shawkins Does it work for you? Or what would you change?

That sounds good. Thank you @mabartos

mabartos commented:
### Progress info

- Removed `SanitizableConfigSources`
- Provided `DisabledMappersInterceptor`
- Provided a workaround for not initializing CMD again 
- General functionality works
- There are some issues with checking if reaugment (the test failures)
- Support duplicated mappers - WiP (80% done) - in different branch - once these issues resolved, it'll be integrated here

mabartos committed:
Fix rendering config options in docs

Fixes #26515

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos commented:
Related bug of the docs: https://github.com/keycloak/keycloak/issues/26515

This resolves the bug.

mabartos commented:
## Potential final changes description
It'd be good to describe and possibly explain all aspects of this PR, as it's quite comprehensive. 
I'd divide it into 3 parts based on present commits:

1. Disabled mappers 
2. Duplicated keys
3. Docs

### 1) Disabled mappers
commit: [Conditionally enable and disable CLI options](https://github.com/keycloak/keycloak/pull/25333/commits/20993e074353df866590f0955a0a4ffea8600843)

- Property mappers can be disabled based on **specified condition**
- When the mapper is disabled, Keycloak does not handle the property - null value is given (as for non-existing)
- Conditions are **lazily evaluated**, so it's possible, during the evaluation, obtaining value even from possibly disabled mapper
- The major part is in the `DisabledMappersInterceptor`, which is a config interceptor handling the disabled mappers
- The `DisabledMappersInterceptor` can be turned OFF 
- The interceptor is turned **OFF** when evaluating the conditions for mappers
- After the evaluation, a **sanitization** of stored mappers **is executed**
- The interceptor is turned **ON** 
- When some disabled option is provided via **CLI**, an **error message is shown**
- When some disabled option is provided via **other source**s, only a **warn log is printed out** with a message that these props are ignored
- More info here: https://github.com/keycloak/keycloak/pull/25333#issuecomment-1893909608

### 2) Duplicated keys
commit: [Support for duplicates in config](https://github.com/keycloak/keycloak/pull/25333/commits/257b2a4b64f08e860649cfe9fce1b2e5d7e1eef3)

We already have some pseudo-support for duplicated keys, but perhaps not intentional as such. Some of the property mappers use the same Option for referencing other stuff - f.e. export and import use `--file` and `--dir` options, which are then mapped to other options. For the import, the `--override` is mapped to specific properties, based on whether `--file` or `--dir` is used. 

-  `PropertyMappers` are represented as `MultivaluedMap<String, PropertyMapper<?>>` for storing several property mappers with the same name
-  PropertyMappers(options) are unique based on their category - one category can contain only one mapper with the same name
- So, it's possible to have `ImportOptions.FILE` and `ExportOptions.FILE` in the application with the same key
- When there are more mappers in the same category, the exception is thrown and developer needs to amend it
- Based on the list of supported categories for the command, we return the found mapper - f.e. for `import` command, we return the `ImportOptions.FILE`, for `export` command, we return the `ExportOptions.FILE`, ...
- When there's some clash of these categories, we return the first found - as in the current Keycloak - options defined as `from` are not transformed, so we should obtain the same result

### 3) Docs
commit: [Fix rendering config options in docs](https://github.com/keycloak/keycloak/pull/25333/commits/9877afdb1bc93a354b3c624ff7ed380070d9a487)
_Be described soon_

shawkins commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java:
```diff
@@ -162,6 +322,29 @@ public Map<OptionCategory, List<PropertyMapper<?>>> getRuntimeMappers() {
+                        final boolean shouldRemove = !pm.isEnabled();
+                        if (shouldRemove) {
+                            MAPPERS.removeMapper(pm);
+                            disabledMappers.put(pm.getTo(), pm);
```
just a nit, we could consider a function to handle this and for the other case - static void addMapper(PropertyMapper mapper, BiConsumer<String, PropertyMapper> adder) ...

mabartos replied:
Done here:
https://github.com/keycloak/keycloak/pull/25333/files#diff-a1bb1ef613469cf669070d52be8a88079005dcaed860e8d11d4231b5e2f530f7R343-R351

shawkins commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/LoggingPropertyMappers.java:
```diff
@@ -29,15 +31,18 @@ public static PropertyMapper<?>[] getMappers() {
                         .paramLabel("<handler>")
                         .build(),
                 fromOption(LoggingOptions.LOG_CONSOLE_OUTPUT)
+                        .isEnabled(LoggingPropertyMappers::isConsoleEnabled, "Console log handler is activated")
```
It would be good to collect the whenEnabled constants to single variables.

mabartos replied:
Done here:
https://github.com/keycloak/keycloak/pull/25333/files#diff-78e1ea4992f437dbda8637131776d9f1e448a2cf4f6ba2dd986e07ec0d1eb298R26-R28

shawkins commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ExportPropertyMappers.java:
```diff
@@ -49,43 +53,70 @@ public static PropertyMapper<?>[] getMappers() {
+                .map(f -> DIR);
+
+        // Only one option can be specified
+        boolean xor = file.isPresent() ^ dir.isPresent();
```
PropertyMappers now also support validators. In this case the actual validation seems to be performed up in the command classes, but may be possible to move it onto the Import/ExportPropertyMappers logic. Not sure that it would really matter - it's just a question of where the logic appears.

mabartos replied:
It's a little bit more complicated. It's not possible to use the present validation logic in PropertyMapper. The thing is, that the ConfigValue in the `validateConfig` is obtained from the `.getFrom()` property, where the transformer is not executed, thus, the value is not computed, thus null. 

The validation is skipped for mappers, where the value is null. 

In order to execute additional validation, even when the value is null, we should do sth like this:
https://github.com/mabartos/keycloak/commit/2b882a302dc61b749c963b0ab8bb9f7ef49f781d

Probably, there might be a better solution, but I'd also rather solve this case in a follow-up task as it's little bit out of scope of this PR. At least, I put the validation logic to the mappers class and referenced from the command validation method. 

@shawkins Does it work for you, to resolve it in a follow-up task?

mabartos replied:
Created follow-up task: https://github.com/keycloak/keycloak/issues/27025

shawkins reviewed (commented):
@mabartos this looks great. Just a couple of thoughts.

mabartos committed:
Conditionally enable and disable CLI options

Closes #13113

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos committed:
Support for duplicates in config

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMappers.java:
```diff
@@ -137,22 +242,73 @@ public void addAll(PropertyMapper<?>[] mappers) {
+
+        public void removeMapper(PropertyMapper<?> mapper) {
+            if (!mapper.getFrom().equals(mapper.getTo())) {
+                remove(mapper.getTo());
```
The intention behind this was that we should not allow duplicates on the "transformed-to" (attribute 'to') properties. Bcs when it happens, we can have a clash of possible transformers for the property, and the result might not be deterministic. 

However, another validation would need to be added for that case, which is not present atm. I think addressing this case could be done in a follow-up task.

So, IMO, it could be handled as you suggested for now.

I'll create the follow-up issues and link them here.

mabartos replied:
Done here: https://github.com/keycloak/keycloak/pull/25333/files#diff-a1bb1ef613469cf669070d52be8a88079005dcaed860e8d11d4231b5e2f530f7R331-R338

shawkins reviewed (dismissed):
LGTM, great work @mabartos

mabartos commented:
@shawkins Thanks! 

I can confirm that everything works as expected, even for the keycloak-benchmark project. @mhajas helped me set up a cluster used for the cross-dc with this PR on top of the Keycloak distro. I think it's a reasonable verification that nothing will break even for them as they're using some additional properties.

vmuzikar reviewed (commented):
@mabartos Thank you for this change (and all the blood, sweat and tears you put into it to make that happen :D). Very nice work!

It was a bit hard to review as it adds a lot of additional complexity. But in general as far as I can tell LGTM.

Do we have any perf benchmark comparison? I know there's a slight impact on the startup performance. Is there any hit on runtime too?

mabartos committed:
Reorder OptionsDistTest

Signed-off-by: Martin Bartoš <mabartos@redhat.com>

mabartos commented:
Rebased and added test execution order for the `OptionsDistTest`. It's not possible to execute `start-dev`, and then `start` without explicit `build` execution. It's the current behavior in Keycloak. We might discuss some improvements around it, but it's out of the scope of this PR.

> It was a bit hard to review as it adds a lot of additional complexity.

@vmuzikar Yep, I'm aware of it. But the code should be cleaner now and also catch some minor issues.

> Do we have any perf benchmark comparison? I know there's a slight impact on the startup performance. Is there any hit on runtime too?

There should not be any performance degradation even for the startup; I resolved that one I mentioned previously. I made only a small perf analysis and nothing that could impact performance :)) The only thing that has changed is that the initialization of Profile(quite resource-consuming) is done earlier.

I can double-check it. 

@shawkins Could you please review the last commit? Thanks!

mabartos commented:
Created follow-up tasks and addressed in the parent issue (https://github.com/keycloak/keycloak/issues/13113):

- [ ] https://github.com/keycloak/keycloak/issues/27025
- [ ] https://github.com/keycloak/keycloak/issues/27026

shawkins reviewed (dismissed):
@mabartos still LGTM. One minor after thought is at what point could or should we be able to clear some of the static state - how big of a footprint does the parsed command have?

mabartos commented:
@shawkins @vmuzikar Can you please check it again? Thanks!

shawkins reviewed (approved):
still looks good!

mabartos commented:
> One minor after thought is at what point could or should we be able to clear some of the static state - how big of a footprint does the parsed command have?

@shawkins Sorry, it seems I've missed your comment. 

I assume you mean the static object `Environment.parsedCommand`. AFAIK pointers/references of static objects are stored in Metaspace, which directs to objects stored in the heap. After getting the heap dump, I noticed that the retained memory for the `Environment.parsedCommand` is ~28.5kB with the inner `CommandSpec` objects. 

However, as the GC was executed before the heap dump, all other `CommandLine#CommandSpecs` were still present, so they're not garbage collected either. That means the parent CommandSpec still references the child specs - I assume it's stored even for the reflection capabilities of Picocli. Thus, IMHO, the footprint does not really matter in this case as there should not be any major differences due to the shared references.

---------------------------------

@vmuzikar May I ask you for a next round of review? Thanks

vmuzikar reviewed (approved):
@mabartos LGTM, fantastic work!

I'll rebase it for the last time to be sure we get accurate test results and enable auto-merge.

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/ShortErrorMessageHandler.java:
```diff
@@ -23,17 +27,30 @@ public int handleParseException(ParameterException ex, String[] args) {
 
-            if (mapper == null || !(cmd.getCommand() instanceof AbstractCommand)) {
+            final boolean isDisabledOption = mapper == null && PropertyMappers.isDisabledMapper(cliKey);
+            final BooleanSupplier isUnknownOption = () -> mapper == null || !(cmd.getCommand() instanceof AbstractCommand);
```
Yep, in this case, it doesn't matter. It's probably some sort of habit to use lazy evaluations everywhere :roll_eyes:

mabartos commented:
@shawkins @radcortez @vmuzikar Thanks for the reviews! :)
